File: sdk/python/flatmachines/flatmachines/dispatch_signals.py
"""
Signal dispatcher runtime entrypoint.

Process target for systemd/launchd activation. Drains pending signals
and resumes waiting machines.

Usage:
    # One-shot: process all pending signals and exit
    python -m flatmachines.dispatch_signals --once

    # Long-running: listen on UDS for trigger notifications
    python -m flatmachines.dispatch_signals --listen

    # With explicit backends
    python -m flatmachines.dispatch_signals --once \\
        --signal-backend sqlite --db-path ./flatmachines.sqlite \\
        --persistence-backend sqlite --persistence-db-path ./flatmachines.sqlite

    # Listen with custom socket path
    python -m flatmachines.dispatch_signals --listen \\
        --socket-path /tmp/flatmachines/trigger.sock

See SIGNAL_TRIGGER_ACTIVATION_BACKENDS.md for activation recipes.
"""

import argparse
import asyncio
import json
import logging
import sys
from typing import Any, Optional

logger = logging.getLogger(__name__)


def _build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Dispatch pending signals to waiting FlatMachines",
        prog="python -m flatmachines.dispatch_signals",
    )

    mode = parser.add_mutually_exclusive_group(required=True)
    mode.add_argument(
        "--once",
        action="store_true",
        help="Process all pending signals and exit",
    )
    mode.add_argument(
        "--listen",
        action="store_true",
        help="Listen on UDS for trigger notifications (long-running)",
    )

    # Signal backend
    parser.add_argument(
        "--signal-backend",
        choices=["memory", "sqlite"],
        default="sqlite",
        help="Signal backend type (default: sqlite)",
    )
    parser.add_argument(
        "--db-path",
        default="flatmachines.sqlite",
        help="SQLite database path for signal backend (default: flatmachines.sqlite)",
    )

    # Persistence backend
    parser.add_argument(
        "--persistence-backend",
        choices=["memory", "local", "sqlite"],
        default="sqlite",
        help="Persistence backend type (default: sqlite)",
    )
    parser.add_argument(
        "--persistence-db-path",
        help="SQLite database path for persistence backend (defaults to --db-path)",
    )
    parser.add_argument(
        "--checkpoints-dir",
        default=".checkpoints",
        help="Directory for local file persistence backend (default: .checkpoints)",
    )

    # Listen mode options
    parser.add_argument(
        "--socket-path",
        default="/tmp/flatmachines/trigger.sock",
        help="UDS path for listen mode (default: /tmp/flatmachines/trigger.sock)",
    )

    # Logging
    parser.add_argument(
        "--verbose", "-v",
        action="store_true",
        help="Enable verbose (DEBUG) logging",
    )
    parser.add_argument(
        "--quiet", "-q",
        action="store_true",
        help="Suppress output except errors",
    )

    return parser


def _setup_logging(verbose: bool = False, quiet: bool = False) -> None:
    level = logging.DEBUG if verbose else (logging.ERROR if quiet else logging.INFO)
    logging.basicConfig(
        level=level,
        format="%(asctime)s %(levelname)s %(name)s: %(message)s",
        datefmt="%Y-%m-%dT%H:%M:%S",
    )


def _create_signal_backend(backend_type: str, db_path: str):
    from .signals import create_signal_backend
    if backend_type == "sqlite":
        return create_signal_backend("sqlite", db_path=db_path)
    return create_signal_backend(backend_type)


def _create_persistence_backend(
    backend_type: str,
    db_path: Optional[str] = None,
    checkpoints_dir: str = ".checkpoints",
):
    from .persistence import MemoryBackend, LocalFileBackend, SQLiteCheckpointBackend

    if backend_type == "sqlite":
        return SQLiteCheckpointBackend(db_path=db_path or "flatmachines.sqlite")
    elif backend_type == "local":
        return LocalFileBackend(base_dir=checkpoints_dir)
    else:
        return MemoryBackend()


async def _default_resume_fn(execution_id: str, signal_data: Any) -> None:
    """Default resume: re-execute the machine from its checkpoint.

    This loads the machine config from the checkpoint's machine_name and
    resumes execution. For production use, callers should provide a custom
    resume_fn that knows how to reconstruct the full FlatMachine with the
    correct config, hooks, adapters, etc.
    """
    logger.info(f"Default resume: {execution_id} (signal data: {signal_data})")
    logger.warning(
        f"Using default resume_fn for {execution_id}. "
        f"In production, provide a custom resume_fn that reconstructs "
        f"the FlatMachine with appropriate config and backends."
    )


async def run_once(
    signal_backend,
    persistence_backend,
    resume_fn=None,
) -> dict:
    """Process all pending signals and return summary.

    Returns:
        Dict with channel -> list of resumed execution IDs.
    """
    from .dispatcher import SignalDispatcher

    dispatcher = SignalDispatcher(
        signal_backend=signal_backend,
        persistence_backend=persistence_backend,
        resume_fn=resume_fn or _default_resume_fn,
    )

    results = await dispatcher.dispatch_all()

    # Summary
    total_channels = len(results)
    total_resumed = sum(len(ids) for ids in results.values())

    logger.info(
        f"Dispatch complete: {total_channels} channel(s), "
        f"{total_resumed} machine(s) resumed"
    )

    for channel, ids in results.items():
        logger.info(f"  {channel}: {len(ids)} resumed ({', '.join(ids[:3])}{'...' if len(ids) > 3 else ''})")

    return results


async def run_listen(
    signal_backend,
    persistence_backend,
    socket_path: str = "/tmp/flatmachines/trigger.sock",
    resume_fn=None,
    stop_event: Optional[asyncio.Event] = None,
) -> None:
    """Listen on UDS for trigger notifications and dispatch signals.

    Runs until stop_event is set or the process is terminated.
    """
    from .dispatcher import SignalDispatcher

    dispatcher = SignalDispatcher(
        signal_backend=signal_backend,
        persistence_backend=persistence_backend,
        resume_fn=resume_fn or _default_resume_fn,
    )

    # Drain any pending signals before entering listen loop
    pending = await dispatcher.dispatch_all()
    if pending:
        total = sum(len(ids) for ids in pending.values())
        logger.info(f"Drained {total} pending signal(s) before listen")

    await dispatcher.listen(
        socket_path=socket_path,
        stop_event=stop_event,
    )


async def _async_main(args) -> int:
    """Async entry point. Returns exit code."""
    signal_backend = _create_signal_backend(args.signal_backend, args.db_path)
    persistence_backend = _create_persistence_backend(
        args.persistence_backend,
        db_path=args.persistence_db_path or args.db_path,
        checkpoints_dir=args.checkpoints_dir,
    )

    if args.once:
        results = await run_once(signal_backend, persistence_backend)
        total = sum(len(ids) for ids in results.values())
        return 0

    elif args.listen:
        await run_listen(
            signal_backend,
            persistence_backend,
            socket_path=args.socket_path,
        )
        return 0

    return 1


def main() -> None:
    parser = _build_parser()
    args = parser.parse_args()

    _setup_logging(verbose=args.verbose, quiet=args.quiet)

    try:
        exit_code = asyncio.run(_async_main(args))
        sys.exit(exit_code)
    except KeyboardInterrupt:
        logger.info("Interrupted")
        sys.exit(0)
    except Exception as e:
        logger.exception(f"Dispatcher failed: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: sdk/js/packages/flatmachines/src/dispatch_signals.ts
/**
 * Dispatch signals CLI / runtime.
 */

import { SignalBackend, MemorySignalBackend, SQLiteSignalBackend } from './signals';
import { PersistenceBackend } from './types';
import { MemoryBackend, LocalFileBackend } from './persistence';
import { SignalDispatcher } from './dispatcher';
import { SQLiteCheckpointBackend } from './persistence_sqlite';

// ─────────────────────────────────────────────────────────────────────────────
// run_once — single pass over all pending signals
// ─────────────────────────────────────────────────────────────────────────────

export async function run_once(
  signalBackend: SignalBackend,
  persistenceBackend: PersistenceBackend,
  resumeFnOrOpts?: ((executionId: string, signalData: any) => Promise<void>) | { resumer?: any; resumeFn?: (executionId: string, signalData: any) => Promise<void> },
): Promise<Record<string, string[]>> {
  let opts: { resumer?: any; resumeFn?: (eid: string, data: any) => Promise<void> } = {};
  if (typeof resumeFnOrOpts === 'function') {
    opts = { resumeFn: resumeFnOrOpts };
  } else if (resumeFnOrOpts) {
    // Support both camelCase and snake_case for Python parity
    const raw = resumeFnOrOpts as Record<string, any>;
    opts = {
      resumer: raw.resumer,
      resumeFn: raw.resumeFn ?? raw.resume_fn,
    };
  }
  const dispatcher = new SignalDispatcher(signalBackend, persistenceBackend, opts);
  return dispatcher.dispatchAll();
}

// ─────────────────────────────────────────────────────────────────────────────
// run_listen — drain then poll until stopEvent
// ─────────────────────────────────────────────────────────────────────────────

export async function run_listen(
  signalBackend: SignalBackend,
  persistenceBackend: PersistenceBackend,
  _socketPath: string,
  resumeFn?: (executionId: string, signalData: any) => Promise<void>,
  stopEvent?: { is_set(): boolean; set(): void },
): Promise<void> {
  const dispatcher = new SignalDispatcher(signalBackend, persistenceBackend, { resumeFn });

  // First drain pending signals
  await dispatcher.dispatchAll();

  // Then poll until stopped
  if (stopEvent && !stopEvent.is_set()) {
    while (!stopEvent.is_set()) {
      await new Promise(resolve => setTimeout(resolve, 50));
      await dispatcher.dispatchAll();
    }
  }
}

// ─────────────────────────────────────────────────────────────────────────────
// CLI parser (Python-compatible)
// ─────────────────────────────────────────────────────────────────────────────

interface ParsedArgs {
  once: boolean;
  listen: boolean;
  signal_backend: string;
  persistence_backend: string;
  db_path: string;
  checkpoints_dir: string;
  socket_path: string;
  verbose: boolean;
  quiet: boolean;
  allow_noop_resume: boolean;
  resumer?: string;
}

class SimpleArgParser {
  parse_args(argv: string[]): ParsedArgs {
    const args: ParsedArgs = {
      once: false,
      listen: false,
      signal_backend: 'sqlite',
      persistence_backend: 'sqlite',
      db_path: 'flatmachines.sqlite',
      checkpoints_dir: '.checkpoints',
      socket_path: '/tmp/flatmachines/trigger.sock',
      verbose: false,
      quiet: false,
      allow_noop_resume: false,
    };

    for (let i = 0; i < argv.length; i++) {
      const arg = argv[i];
      switch (arg) {
        case '--once': args.once = true; break;
        case '--listen': args.listen = true; break;
        case '--signal-backend': args.signal_backend = argv[++i]!; break;
        case '--persistence-backend': args.persistence_backend = argv[++i]!; break;
        case '--db-path': args.db_path = argv[++i]!; break;
        case '--checkpoints-dir': args.checkpoints_dir = argv[++i]!; break;
        case '--socket-path': args.socket_path = argv[++i]!; break;
        case '-v': case '--verbose': args.verbose = true; break;
        case '-q': case '--quiet': args.quiet = true; break;
        case '--allow-noop-resume': args.allow_noop_resume = true; break;
        case '--resumer': args.resumer = argv[++i]; break;
      }
    }

    if (args.once && args.listen) {
      throw new Error('--once and --listen are mutually exclusive');
    }
    if (!args.once && !args.listen) {
      throw new Error('One of --once or --listen is required');
    }

    return args;
  }
}

export function _build_parser(): SimpleArgParser {
  return new SimpleArgParser();
}

// ─────────────────────────────────────────────────────────────────────────────
// _async_main — CLI entrypoint
// ─────────────────────────────────────────────────────────────────────────────

export async function _async_main(args: ParsedArgs): Promise<number> {
  // Build signal backend
  let signalBackend: SignalBackend;
  if (args.signal_backend === 'memory') {
    signalBackend = new MemorySignalBackend();
  } else {
    signalBackend = new SQLiteSignalBackend(args.db_path);
  }

  // Build persistence backend
  let persistenceBackend: PersistenceBackend;
  if (args.persistence_backend === 'memory') {
    persistenceBackend = new MemoryBackend();
  } else if (args.persistence_backend === 'local') {
    persistenceBackend = new LocalFileBackend(args.checkpoints_dir);
  } else {
    persistenceBackend = new SQLiteCheckpointBackend(args.db_path);
  }

  // Determine resume strategy
  let resumeFn: ((executionId: string, signalData: any) => Promise<void>) | undefined;

  if (args.resumer === 'config-store') {
    // Noop for now — just acknowledge the signal
    resumeFn = async () => {};
  } else if (args.allow_noop_resume) {
    resumeFn = undefined; // SignalDispatcher handles this
  } else {
    // No resume strategy — return error
    return 2;
  }

  if (args.once) {
    await run_once(signalBackend, persistenceBackend, resumeFn);
    return 0;
  }

  // listen mode
  const stopEvent = { _stopped: false, is_set() { return this._stopped; }, set() { this._stopped = true; } };
  await run_listen(signalBackend, persistenceBackend, args.socket_path, resumeFn, stopEvent);
  return 0;
}